### PR TITLE
feat: Add TLS support to gnet

### LINKED ISSUE #16: TLS support?
Is TLS support planned?

In accordance with https://github.com/tidwall/evio/issues/28 there are no easy way to implement it:
> The problem is in standard Go TLS implementation. It doesn't have an event-based idea mechanism and actually it can block calls like Reader.Read() during handshake, what is actually not accurate by the io.Reader contract.

Looks like custom TLS implementation is required for this.

### PR DESCRIPTION
**Pull request by 0-haha**
## 1. Are you opening this pull request for bug-fixes, optimizations or new feature?

new feature

## 2. Please describe how these code changes achieve your intention.

This PR is to add the TLS support to get.

#### Change of the source code
- The main TLS library is packaged in the directory `pkg/tls/`
- `internal/boring` is the dummy package for `boring TLS` as it is required by the standard Golang TLS library
- Other changes go to `acceptor.go`, `connection.go`, and `eventloop.go`. Basically, once the TLS enabled on the server side,
   it will call `gnetConn.UpgradeTLS()` to upgrade the protocol to TLS. Then, all reads and writes will go to the `gnetConn.readTLS()` and `gnetConn.writeTLS()`

#### The gnet TLS implementation
- is developed based on https://github.com/luyu6056/tls.
- merges the upstream go v1.20rc3 standard TLS library.
    - Since go 1.20 uses `crypto/ecdh` in `crypto/tls/key_agreement.go`, which is not available in `go <= 1.19`, `go.mod` is bumped up to `1.20`.
- adds the Kernel TLS support. So one can offload the encryption to the kernel.
    - The implementation of kernel TLS was based on [@jim3m](https://github.com/jim3ma/go/tree/dev.ktls.1.16.3)'s implementation,  which was based on [@FiloSottile](https://github.com/FiloSottile/go/commit/dbed9972d9947eb0001e9f5b639e0df05acec8bd)'s implementation.
    - Kernel TLS is totally depending on the kernel version.
    - Kernel TLS Features
        - KTLS 1.2 TX & RX
        - KTLS 1.3 TX & RX
        - zerocopy and no pad for TLS 1.3
        - ciphersuites: AES-GCM-128, AES-GCM-256, CHACHA20POLY1305 
   - Kernel TLS TODO
       - KTLS 1.3 RX disabled on kernel < 5.19 as it causes weird package lost
       - zero copy and no pad have not been tested yet. zero copy is enabled on `kernel >= 5.19`, and no pad is enabled on `kernel >= 6.0`.
       - sendfile api

#### Examples
An example of using gnet TLS can found at https://github.com/0-haha/gnet_tls_examples. You should be able to run the repo in docker.

#### Other Comments
I would say the majority of the implementation has been completed. Open to ongoing conversations.
中文可以聊。

## 3. Please link to the relevant issues (if any).

Fixes #16 

## 4. Which documentation changes (if any) need to be made/updated because of this PR?

The `gnet.Run` command will accept the `tls.Config` like this:

```go
cer, _ := tls.LoadX509KeyPair("server.crt", "server.key")
// server only uses TLS 1.2 and TLS 1.3
config := &tls.Config{
    MinVersion:   tls.VersionTLS12,
    Certificates: []tls.Certificate{cer},
}
gnet.Run(echo, "tcp://192.168.0.100:8000", gnet.WithTLS(config))
```

## 4. Checklist

- [ x] I have squashed all insignificant commits.
- [ x] I have commented my code for explaining package types, values, functions, and non-obvious lines.
- [ ] I have written unit tests and verified that all tests passes (if needed).
- [ ] I have documented feature info on the README (only when this PR is adding a new feature).
- [ x] (optional) I am willing to help maintain this change if there are issues with it later.

### CODE REVIEW

**Comment by panjf2000** [comment]
Thank you for implementing TLS and opening this PR.

This might cost me a lot of time to absorb and review the code, but I'll do this as fast as possible.

**Comment by 0-haha** [comment]
I optimized the memory copy and buffer usage for TLS read, write, and handshake. The following briefly describes the implementation idea which would be helpful to review the code

## Buffers used in TLS
- `rawInput`: stores the TLS record from TCP
- `input`: stores the decrypted TLS record, and the memory is owned by `rawInput`
- `hand`: stores the handshake data
- `sendBuf`: stores the sending data, and it is only used during the handshake

## TLS handshake (starts in `eventloop.read()`):
1. Attach `eventloop.buffer` to `tlsconn.rawInput` (zero-copy)
2. Call the `tlsconn.handshake()`
3. If `tlsconn.HandshakeComplete()`, call `eventloop.readTLS()`; Otherwise, return and will restart at `1` in the next round

Note: In `tlsconn.handshake()`, it extracts the handshake message from `tlsconn.rawInput` and zero-copys the message to `tls.hand`.
`tls.hand` discards the messages immediately after using it.

## TLS read:
1. Attach `eventloop.buffer` to `tlsconn.rawInput` (zero-copy)
2. Call eventloop.readTLS(). Since `tlsconn.rawInput` can holds multiple TLS records, we iteratively process all TLS records.
    ```
    for {
        Extract the TLS record from "tlsconn.rawInput"
        Decrypt the record into "tlsconn.rawInput"
        tlsconn.data = decrypted record (store the reference, zero-copy)
        gnetConn.buffer = tlsconn.Data() (return tlsconn.data, zero-copy)
        eventHandler.OnTraffic()
        c.inboundBuffer.Write(c.buffer)
        if no more TLS records in "tlsconn.rawInput" {
            discard the data in "tlsconn.rawInput" and "tlsconn.data"
            if there is data left in "tlsconn.rawInput", we cache it. so, the left data is not owned by "eventloop.buffer" which will be used by another "gnetConn"
        }
    }
    ```
    The data pipeline in each iteration is
    `eventloop.buffer` (TLS records from TCP) -> (zero-copy) -> `tlsconn.rawInput` -> (decrypt TLS records) -> `tlsconn.rawInput` -> (zero-copy) -> `tlsconn.Data`
    -> (zero-copy) -> `gnetConn.buffer` -> (used by `eventHandler.OnTraffic()`) -> (write rest into to) -> `c.inboundBuffer`

- Kernel TLS read pipeline
  `eventloop.buffer` -> (attach to, zero-copy) -> `tlsconn.rawInput`
  ktlsReadRecord -> (decrypt) -> `tlsconn.rawInput` (referring to  `eventloop.buffer`) -> (zero-copy) -> `tlsconn.Data`
 
  The rest follow the standard gnet TLS read.

## TLS write:
The data pipeline is
`data` -> (encrypt) -> `buffer` from `sync.Pool` -> (call) -> `gnetConn.writeTCP()` -> (call) ->  `gnetConn.write()` (standard gnet write function) -> return `buffer` to`sync.Pool`

- Kernel TLS write
  call `gnetConn.write(data)`. When `gnetConn` writes the data into the socket (call `unix.Write()`), kernel encrypts the data automatically.

**Comment by 0-haha** [comment]
Merge Go upstream commits

**Comment by mostafa** [comment]
Hey @0-haha @panjf2000,

If you ever need help testing this, I'd be very happy to help.

**Comment by panjf2000** [comment]
> Hey @0-haha @panjf2000,
> 
> If you ever need help testing this, I'd be very happy to help.

Thank you for offering to help, I'll reach out to you in case needed.

**Comment by panjf2000** [comment]
A new thought about TLS in `gnet`: is it possible to implement TLS as an external library that we are able to plug into `gnet` (or offload it from `gnet`) easily?

**Comment by 0-haha** [comment]
> A new thought about TLS in `gnet`: is it possible to implement TLS as an external library that we are able to plug into `gnet` (or offload it from `gnet`) easily?

Actually, I have the same idea, like what the quic library did. See https://github.com/quic-go/quic-go/tree/master/internal/qtls

Then, it can support multiple go versions at the same time.

**Comment by 0-haha** [comment]
@panjf2000 The TLS implementation has become an external library. `gnet` can support multiple TLS versions (each binding to a Go version).

See if you are happy with that.

**Comment by panjf2000** [comment]
> @panjf2000 The TLS implementation has become an external library. `gnet` can support multiple TLS versions (each binding to a Go version).
> 
> See if you are happy with that.

Thank you so much for the efforts, I'll review this PR in the next few days.

File: connection.go [hunk 1]
```diff
@@ -127,8 +129,18 @@ func (c *conn) open(buf []byte) error {
 	return err
 }
 
+func (c *conn) writeTLS(data []byte) (n int, err error) {
+	// use tls to encrypt the data before sending it.
+	// tlsconn will call gnet.WriteTCP() to sent the data directly.
+	// If gnetConn.outboundBufferis not empty, data will be
+	// buffered in gnetConn.outboundBuffer.
+	n, err = c.tlsconn.Write(data)
```
Comment (panjf2000) [inline connection.go:137]: Why don't you use `outboundBuffer` here?

File: connection.go [hunk 2]
```diff
@@ -154,6 +166,28 @@ func (c *conn) write(data []byte) (n int, err error) {
...
+	}
+
+	// use tls to encrypt the data before sending it.
+	// tlsconn will call gnet.WriteTCP() to sent the data directly.
+	// If gnetConn.outboundBufferis not empty, data will be
+	// buffered in gnetConn.outboundBuffer.
+	sent := 0
+	var sentN int
+	for _, b := range bs {
+		sentN, err = c.tlsconn.Write(b)
```
Comment (panjf2000) [inline connection.go:181]: ditto

File: connection.go [hunk 3]
```diff
@@ -127,8 +129,18 @@ func (c *conn) open(buf []byte) error {
 	return err
 }
 
+func (c *conn) writeTLS(data []byte) (n int, err error) {
+	// use tls to encrypt the data before sending it.
+	// tlsconn will call gnet.WriteTCP() to sent the data directly.
+	// If gnetConn.outboundBufferis not empty, data will be
```
Comment (panjf2000) [inline connection.go:135]: gnetConn.outboundBufferis --> gnet.Conn.outboundBuffer is

File: connection.go [hunk 4]
```diff
@@ -348,13 +390,25 @@ func (c *conn) Write(p []byte) (int, error) {
...
 	}
+	if c.tlsconn != nil {
+		return c.writeTLS(p)
+	}
+	return c.write(p)
+}
+
+// Expose the plaintext write API which should only be used
+// by tlsconn.Write().
+func (c *conn) WriteTCP(p []byte) (int, error) {
```
Comment (panjf2000) [inline connection.go:401]: Why do we need this new method?

File: connection.go [hunk 5]
```diff
@@ -467,3 +521,27 @@ func (c *conn) Close() error {
 		return
 	}, nil)
 }
+
+func (c *conn) UpgradeTLS(config *tls.Config) (err error) {
+	// TODO: create a sync.pool to manage the TLS connection
+	c.tlsconn = tls.Server(c, config.Clone())
+
+	// 很有可能握手包在UpgradeTls之前发过来了，这里把inboundBuffer剩余数据当做握手数据处理
```
Comment (panjf2000) [inline connection.go:529]: 注释改成英文。

File: eventloop.go [hunk 6]
```diff
@@ -126,6 +177,29 @@ func (el *eventloop) read(c *conn) error {
 		return el.closeConn(c, os.NewSyscallError("read", err))
 	}
 
+	if c.tlsconn != nil {
+		// attach the gnet eventloop.buffer to tlsconn.rawInput.
+		c.tlsconn.RawInputSet(el.buffer[:n]) //nolint:errcheck
+		if !c.tlsconn.HandshakeComplete() {
+			// 先判断是否足够一条消息
```
Comment (panjf2000) [inline eventloop.go:184]: 注释改成英文

File: go.mod [hunk 7]
```diff
@@ -1,11 +1,12 @@
 module github.com/panjf2000/gnet/v2
 
 require (
+	github.com/0-haha/gnet_go_tls/v120 v120.2.0
```
Comment (panjf2000) [inline go.mod:4]: 为什么版本号不是 v1.20.2 ？

**Review by panjf2000 — CHANGES_REQUESTED** [review]

File: go.mod (hunk 7 above)
Comment (0-haha) [reply go.mod:4]: 版本号用 v1.20.2 `go mod tidy` 会报错

```
go: errors parsing go.mod:
/workspace/go.mod:4:2: require github.com/0-haha/gnet_go_tls/v120: version "v1.20.2" invalid: should be v120, not v1
```

为什么用v120 ？

go mod 不支持v1.x, 必须是int。为了支持多版本go，比如未来的v1.21，所以go 1.20的 go mod 链接用 `github.com/0-haha/gnet_go_tls/v120`

File: connection.go [hunk 8]
```diff
@@ -127,8 +129,18 @@ func (c *conn) open(buf []byte) error {
 	return err
 }
 
+func (c *conn) writeTLS(data []byte) (n int, err error) {
+	// use tls to encrypt the data before sending it.
+	// tlsconn will call gnet.WriteTCP() to sent the data directly.
+	// If gnetConn.outboundBufferis not empty, data will be
+	// buffered in gnetConn.outboundBuffer.
```
Comment (0-haha) [inline connection.go:136]: 为什么不用outboundBuffer?

`tlsconn.write` 直接加密数据，然后往socket写数据。如果用outboundBuffer就会多义词copy影响性能。
在kernel tls开启的时候可以，做到zero-copy

Comment (0-haha) [reply connection.go:136]: 手滑点了 resolve conversation

File: connection.go (hunk 1 above)
Comment (0-haha) [reply connection.go:137]: 为什么不用outboundBuffer?

tlsconn.write 直接加密数据，然后往socket写数据。如果用outboundBuffer就会多义词copy影响性能。 在kernel tls开启的时候可以，做到zero-copy

File: go.mod (hunk 7 above)
Comment (panjf2000) [reply go.mod:4]: 你这种 go mod 版本号的用法太奇怪了，我理解并不需要在 go.mod 里加上 v120，后面的版本号通过 github release 和 tag 来管理就行了。

Comment (0-haha) [reply go.mod:4]: 这个是针对go 1.20的实现。
https://github.com/0-haha/gnet/blob/dev/pkg/tls/go120.go

针对不同go 版本用go:build 来区分编译的包

go-quic 是直接每个不同的go 版本一个git repo。你看一下他的import就知道了。
https://github.com/quic-go/quic-go/tree/master/internal/qtls

目前还没想到更好的办法在一个仓库里维护多个版本的办法

File: connection.go (hunk 5 above)
Comment (0-haha) [reply connection.go]: Resolved in new commits

File: eventloop.go (hunk 6 above)
Comment (0-haha) [reply eventloop.go]: Resolved in new commits

File: connection.go (hunk 2 above)
Comment (0-haha) [reply connection.go:181]: Resolved in new commits

File: connection.go (hunk 3 above)
Comment (0-haha) [reply connection.go]: Resolved in new commits

File: connection.go (hunk 4 above)
Comment (0-haha) [reply connection.go:401]: tlsconn.Write() 内部最终会向 gnetConn 写数据，但是gnetConn 内部又有tlsconn 的指针，所以就会有一个死循环。
参照 https://github.com/0-haha/gnet_go_tls/blob/5728fd829790624e21452e217702b9c9964ded45/conn.go#L920-L940

因为tlsconn.Write()是把加密后的数据写到gnetconn，所以需要gnetConn暴露只写明文的API就是gnetConn.write()

File: connection.go (hunk 1 above)
Comment (0-haha) [reply connection.go:137]: 因为 tlsconn.Write 会透明加密 data，然后调用gnetConn.WriteTCP() -> gnetConn.write().

这样就可以直接往socket写数据，而不是copy到outboundbuffer，然后再写数据，性能更好。

在kernel tls开启的时候，可以做到zero-copy，因为内核直接透明加密。

Comment (0-haha) [reply connection.go:137]: 因为 tlsconn.Write 会透明加密 data，然后调用gnetConn.WriteTCP() -> gnetConn.write().

这样就可以直接往socket写数据，而不是copy到outboundbuffer，然后再写数据，性能更好。

在kernel tls开启的时候，可以做到zero-copy，因为内核直接透明加密。

File: go.mod (hunk 7 above)
Comment (0-haha) [reply go.mod:4]: @panjf2000  go mod 里面是准备一个go版本一个TLS库吗? 维护就只能fork + cherry pick了。

```
require (
	github.com/0-haha/gnet_go_tls-1-20 v1.20.0
	github.com/0-haha/gnet_go_tls-1-21 v1.21.0
)
```

或者就是一个repo下两个文件夹，一个文件夹对应一个go版本

我就想到这两种方案，其他的试过了，但是使用的时候必须要在go mod里用replace ，直接跑go mod tidy 会报错。

你倾向哪个方案

**Comment by Fgaoxing** [comment]
@panjf2000 请问gnet何时可以支持TLS

**Comment by ruimya** [comment]
@panjf2000 hope to further test this library and make gnet officially support tls as soon as possible.

maybe this library https://github.com/[0-haha/gnet-tls-go1-20](https://github.com/0-haha/gnet-tls-go1-20) should be merged into gnet-io?

**Comment by panjf2000** [comment]
> @panjf2000 hope to further test this library and make gnet officially support tls as soon as possible.
> 
> maybe this library [gnet-tls-go1-20](https://github.com/0-haha/gnet-tls-go1-20) should be merged into gnet-io?

The owner hasn't responded to my code review for a long time, and this PR has been falling the GitHub CI. Therefore, I've decided to close this PR, but for anyone who wants to implement TLS on `gnet`, legit PRs are always welcome!

**Pull request closed without merging**